matscipy/calculators/pair_potential/calculator.py

```diff
@@ -0,0 +1,422 @@
+                mask2 = atnums[j_n] == pair[0]
+                mask = np.logical_and(mask1, mask2)
+
+                e_n[mask] = self.f[pair](abs_dr_n[mask])
+                de_n[mask] = self.df[pair](abs_dr_n[mask])
+
+            if pair[0] != pair[1]:
+                mask1 = np.logical_and(atnums[i_n] == pair[0], atnums[j_n] == pair[1])
+                mask2 = np.logical_and(atnums[i_n] == pair[1], atnums[j_n] == pair[0])
+                mask = np.logical_or(mask1, mask2)
+
+                e_n[mask] = self.f[pair](abs_dr_n[mask])
+                de_n[mask] = self.df[pair](abs_dr_n[mask]) 
+
+        epot = 0.5*np.sum(e_n)
+
+        # Forces
+        df_nc = -0.5*de_n.reshape(-1,1)*dr_nc/abs_dr_n.reshape(-1,1)
+
+        # Sum for each atom
+        fx_i = np.bincount(j_n, weights=df_nc[:,0], minlength=nat) - \
+            np.bincount(i_n, weights=df_nc[:,0], minlength=nat)
+        fy_i = np.bincount(j_n, weights=df_nc[:,1], minlength=nat) - \
+            np.bincount(i_n, weights=df_nc[:,1], minlength=nat)
+        fz_i = np.bincount(j_n, weights=df_nc[:,2], minlength=nat) - \
+            np.bincount(i_n, weights=df_nc[:,2], minlength=nat)
+
+        # Virial
+        virial_v = -np.array([dr_nc[:,0]*df_nc[:,0],               # xx
+                              dr_nc[:,1]*df_nc[:,1],               # yy
+                              dr_nc[:,2]*df_nc[:,2],               # zz
+                              dr_nc[:,1]*df_nc[:,2],               # yz
+                              dr_nc[:,0]*df_nc[:,2],               # xz
+                              dr_nc[:,0]*df_nc[:,1]]).sum(axis=1)  # xy
+
+
+        self.results = {'energy': epot,
+                        'stress': virial_v/self.atoms.get_volume(),
+                        'forces': np.transpose([fx_i, fy_i, fz_i])}
+
+    ###
+
+    def calculate_hessian_matrix(self, atoms, H_format="dense"):
+        """
+        Calculate the Hessian matrix for a pair potential.
+        For an atomic configuration with N atoms in d dimensions the hessian matrix is a symmetric, hermitian matrix 
+        with a shape of (d*N,d*N). The matrix is in general a sparse matrix, which consists of dense blocks of shape (d,d), which
+        are the mixed second derivatives. The result of the derivation for a pair potential can be found in:
+        L. Pastewka et. al. "Seamless elastic boundaries for atomistic calculations", Phys. Ev. B 86, 075459 (2012)
+
+        Parameters
+        ----------
+        atoms: ase.Atoms
+            Atomic configuration in a local or global minima.
+
+        H_format: "dense" or "sparse"
+            Output format of the hessian matrix. 
+            The format "sparse" is only possible if matscipy was build with scipy.
+        """
+
+        if H_format == "sparse":
+            try:
+                from scipy.sparse import bsr_matrix
+            except:
+                print("Changing format to dense since scipy.sparse could not be loaded!")
```

Comment:
Don't change to dense automatically, just throw an error that scipy is not available. If the users asks for a certain return format she/he should get it (or an error).